### PR TITLE
[WIP] Add separate call for updating DC status

### PR DESCRIPTION
**Pull request by kargakis**
**Blocked on #6337**

@ironcladlou @deads2k 

Needs a couple of unit tests.

Fixes https://github.com/openshift/origin/issues/6465

### CODE REVIEW

**Comment by kargakis** [comment]
`oc deploy --latest` works fine: http://pastebin.com/ynGZvrfR

File: pkg/cmd/cli/cmd/deploy.go [hunk 1]
```diff
@@ -222,7 +222,7 @@ func (o DeployOptions) deploy(config *deployapi.DeploymentConfig, out io.Writer)
 	}
 
 	config.Status.LatestVersion++
-	_, err = o.osClient.DeploymentConfigs(config.Namespace).Update(config)
+	_, err = o.osClient.DeploymentConfigs(config.Namespace).UpdateStatus(config)
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/deploy.go]: is this backwards compatible?

Comment (liggitt) [inline pkg/cmd/cli/cmd/deploy.go]: actually, we would never want a client to modify status to trigger a deployment... a client would modify spec, the controller would modify status to indicate the requested deployment was done

File: pkg/deploy/registry/deployconfig/strategy.go [hunk 2]
```diff
@@ -64,6 +65,24 @@ func (strategy) CheckGracefulDelete(obj runtime.Object, options *kapi.DeleteOpti
...
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update of status.
+func (statusStrategy) PrepareForUpdate(obj, old runtime.Object) {
+	newDc := obj.(*api.DeploymentConfig)
+	oldDc := old.(*api.DeploymentConfig)
+	newDc.Spec = oldDc.Spec
+}
+
+// ValidateUpdate is the default update validation for an end user updating status.
+func (statusStrategy) ValidateUpdate(ctx kapi.Context, obj, old runtime.Object) fielderrors.ValidationErrorList {
+	return validation.ValidateDeploymentConfigUpdate(obj.(*api.DeploymentConfig), old.(*api.DeploymentConfig))
```
Comment (liggitt) [inline pkg/deploy/registry/deployconfig/strategy.go]: I think I want a separate ValidateDeploymentConfigStatusUpdate method... I don't think we want invalid things in the spec to make status updates fail

File: pkg/deploy/registry/deployconfig/strategy.go [hunk 3]
```diff
@@ -64,6 +65,24 @@ func (strategy) CheckGracefulDelete(obj runtime.Object, options *kapi.DeleteOpti
...
+	strategy
+}
+
+var StatusStrategy = statusStrategy{Strategy}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update of status.
+func (statusStrategy) PrepareForUpdate(obj, old runtime.Object) {
+	newDc := obj.(*api.DeploymentConfig)
+	oldDc := old.(*api.DeploymentConfig)
+	newDc.Spec = oldDc.Spec
```
Comment (liggitt) [inline pkg/deploy/registry/deployconfig/strategy.go]: we allow changing labels and annotations on status update?

File: pkg/cmd/cli/cmd/deploy.go (hunk 1 above)
Comment (kargakis) [inline pkg/cmd/cli/cmd/deploy.go]: There is https://github.com/openshift/origin/issues/6337 for fixing that.

File: pkg/deploy/registry/deployconfig/strategy.go (hunk 3 above)
Comment (kargakis) [inline pkg/deploy/registry/deployconfig/strategy.go]: Hm, nice catch. This needs to be fixed upstream as well.

File: pkg/deploy/registry/deployconfig/strategy.go (hunk 2 above)
Comment (kargakis) [inline pkg/deploy/registry/deployconfig/strategy.go]: Right, will add a separate validation path for status.

File: pkg/cmd/cli/cmd/deploy.go (hunk 1 above)
Comment (liggitt) [inline pkg/cmd/cli/cmd/deploy.go]: I don't want a version of the CLI to exist which makes a call to the `status` subresource in order to trigger a deployment

Comment (kargakis) [inline pkg/cmd/cli/cmd/deploy.go]: I see. Then we need #6337 to be fixed first.

Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: FWIW in the short term, updates to the status won't result in a new deployment unless `status.latestVersion` changes.

**Comment by kargakis** [comment]
Superceded by https://github.com/openshift/origin/pull/7149

**Pull request closed without merging**